x/ibc/02-client/types/tendermint/evidence.go

```diff
@@ -0,0 +1,101 @@
+package tendermint
+
+import (
+	"fmt"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/tendermint/tendermint/crypto/tmhash"
+	cmn "github.com/tendermint/tendermint/libs/common"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	evidenceexported "github.com/cosmos/cosmos-sdk/x/evidence/exported"
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types/errors"
+)
+
+var _ evidenceexported.Evidence = Evidence{}
+
+// Evidence is a wrapper over tendermint's DuplicateVoteEvidence
+// that implements Evidence interface expected by ICS-02
+type Evidence struct {
+	Header1 Header
+	Header2 Header
+	ChainID string
+}
+
+// Route implements Evidence interface
+func (ev Evidence) Route() string {
+	return "client"
+}
+
+// Type implements Evidence interface
+func (ev Evidence) Type() string {
+	return "client_misbehaviour"
+}
+
+// String implements Evidence interface
+func (ev Evidence) String() string {
+	bz, err := yaml.Marshal(ev)
+	if err != nil {
+		panic(err)
+	}
+	return string(bz)
+}
+
+// Hash implements Evidence interface
+func (ev Evidence) Hash() cmn.HexBytes {
+	return tmhash.Sum(SubModuleCdc.MustMarshalBinaryBare(ev))
+}
+
+// GetHeight returns the height at which misbehaviour occurred
+func (ev Evidence) GetHeight() int64 {
+	return ev.Header1.Height
+}
+
+// ValidateBasic implements Evidence interface
+func (ev Evidence) ValidateBasic() error {
+	// ValidateBasic on both validators
+	if err := ev.Header1.ValidateBasic(ev.ChainID); err != nil {
+		return errors.ErrInvalidEvidence(errors.DefaultCodespace, err.Error())
+	}
+	if err := ev.Header2.ValidateBasic(ev.ChainID); err != nil {
+		return errors.ErrInvalidEvidence(errors.DefaultCodespace, err.Error())
+	}
+	// Ensure that Heights are the same
+	if ev.Header1.Height != ev.Header2.Height {
+		return errors.ErrInvalidEvidence(errors.DefaultCodespace, "headers in evidence are on different heights")
+	}
+	// Ensure that Commit Hashes are different
+	if ev.Header1.Commit.BlockID.Equals(ev.Header2.Commit.BlockID) {
+		return errors.ErrInvalidEvidence(errors.DefaultCodespace, "Headers commit to same blockID")
+	}
+
+	if err1 := ValidCommit(ev.ChainID, ev.Header1.Commit, ev.Header1.ValidatorSet); err1 != nil {
+		return err1
+	}
+	if err2 := ValidCommit(ev.ChainID, ev.Header2.Commit, ev.Header2.ValidatorSet); err2 != nil {
+		return err2
+	}
+
+	return nil
+}
+
+func ValidCommit(chainID string, commit *tmtypes.Commit, valSet *tmtypes.ValidatorSet) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
and if so, why are we returning an error instead of sticking with the panic?

Reply:
The `CommitToVoteSet` function can panic. This makes sense in the context of how tendermint uses it, we should return error instead here since it means that someone tried to submit Evidence that wasn't actually committed by validators

This should result in just rejecting the evidence gracefully, not crashing the node

Reply:
Makes sense. Mind adding a comment on the code with that clarification?